README.md
# qccrs: Quality Control Charts under Repetitive Sampling

## Introduction
 The **qccrs** package provides functionalities to calculate  Average Sample Numbers (ASN), Average Run Length (ARL1) and  value of k, k1 and k2 for quality control charts under repetitive sampling as given in Aslam et al. (2014).

## Authors
1. Muhammad Yaseen (myaseen208@gmail.com)
2. Muhammad Aslam  (aslam_ravian@hotmail.com)
3. Sami Ullah      (samiullahuos@gmail.com)
4. Muhammad Azam   (mazam@uvas.edu.pk)
5. Chi-Hyuck Jun   (chjun@postech.ac.kr)
6. Muhammad Kashif (mkashif@uaf.edu.pk)

## Installation
Use **devtools** to install the development version from Github:

```{r}
if(!require("devtools")) install.packages("devtools")
devtools::install_github("myaseen208/qccrs")
```

## License
This package is free and open source software, licensed under GPL.


NEWS.md
# Changes in qccrs Version 0.1.0
## New Features 

* First version of qccrs 


R/xrs.R
#' @name    xrs
#' @aliases xrs
#' @title   Xbar Control Charts Under Repetitive Sampling
#' @description Calculates the Average Sample Number and Average Run Length as given in Aslam et al. (2014)
#'
#' @param .n     Sample Size
#' @param .c     Size of the Shift      
#' @param .k1    Positive Integer
#' @param .k2    Positive Integer
#'
#'
#' @return  Average Sample Number (ASN) and Average Run Length (ARL1) for xbar control charts under repetitive sampling
#'
#' @author
#' \enumerate{
#'          \item Muhammad Yaseen (\email{myaseen208@@gmail.com})
#'          \item Muhammad Aslam (\email{aslam_ravian@@hotmail.com})
#'          \item Sami Ullah (\email{samiullahuos@@gmail.com})
#'          \item Muhammad Azam   (\email{mazam@@uvas.edu.pk})
#'          \item Chi-Hyuck Jun   (\email{chjun@@postech.ac.kr})
#'          \item Muhammad Kashif (\email{mkashif@@uaf.edu.pk})
#'          }
#'
#' @references
#' Aslam, M., Azam, M. and Jun, C. (2014).
#'  New Attributes and Variables Control Charts under Repetitive Sampling.
#'  \emph{Industrial Engineering & Management Systems}.
#'  \strong{1}(13):101-106.
#'
#' @importFrom stats pnorm 
#' @importFrom tibble tibble
#' @importFrom magrittr %>%
#' @importFrom purrr map
#'
#' @export
#'
#' @examples
#' 
#' library(magrittr)
#' library(purrr)
#' 
#' c(0.0, 0.1, 0.20, 0.3, 0.4, 0.5, 1.0, 1.5, 2, 3) %>%
#' purrr::map(
#' function(x) 
#'     xrs(
#'         .c     = x
#'       , .n     = 10
#'       , .k1    = 2.9301
#'       , .k2    = 0.9825))
#'
xrs <- function(.c, .n,  .k1 , .k2 ){
  UseMethod("xrs")
}

#' @export
#' @rdname xrs
xrs.default <- function(.c, .n,  .k1 , .k2 ){
  df <-  tibble::tibble(
       .c     = .c  
      , Pin   = (2* pnorm(q = .k2, mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-1)/
              1- 2*(pnorm(q = .k1, mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-
                    pnorm(q = .k2, mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE))
      , Pinc  = (pnorm(q = .k2 - .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)+
                 pnorm(q = .k2 + .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-1)/
                (pnorm(q = .k2 + .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-
                 pnorm(q = .k1 + .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-
                 pnorm(q = .k1 - .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)+
                 pnorm(q = .k2 - .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)+1)
      , Prep  = 2*(pnorm(q = .k1, mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-
                   pnorm(q = .k2, mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE))
      , Prep1 = -pnorm(q = .k2 + .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)+
                 pnorm(q = .k1 + .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)+
                 pnorm(q = .k1 - .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)-
                 pnorm(q = .k2 - .c*sqrt(.n), mean =0, sd = 1, lower.tail = TRUE, log.p = FALSE)
      , ARL0  =   1/(1-Pin)
      , ARL1  =   1/(1-Pinc)
      , ASN  =   .n/(1-Prep1)
    ) %>%
      dplyr::select(.c, ASN, ARL1)
    return(df)
  }


R/npcrs2.R
#' @name    npcrs2
#' @aliases npcrs2
#' @title   Attributes Control Charts under Repetitive Sampling with two positive integers
#' @description Calculates Average Sample Numbers (ASN), Average Run Length (ARL1) and value of k1 and k2 for attributes control charts under repetitive sampling as given in Aslam et al.(2014)
#'
#' @param .p0     probability that process is in control
#' @param .f      Size of the Shift
#' @param .n      Sample Size
#' @param .ssize  Number of samples with replacement at each iteration
#' @param .k1r    Random postive constant
#' @param .k2r    Random postive constant
#' @param .k1     Fixed positive constant
#' @param .k2     Fixed positive constant
#'
#'
#'
#' @return ASN, ARL, K1 and K2
#'
#' @author
#' \enumerate{
#'          \item Muhammad Yaseen (\email{myaseen208@@gmail.com})
#'          \item Muhammad Aslam  (\email{aslam_ravian@@hotmail.com})
#'          \item Sami Ullah      (\email{samiullahuos@@gmail.com})
#'          \item Muhammad Azam   (\email{mazam@@uvas.edu.pk})
#'          \item Chi-Hyuck Jun   (\email{chjun@@postech.ac.kr})
#'          \item Muhammad Kashif (\email{mkashif@@uaf.edu.pk})
#'          }
#'
#' @references
#' Aslam, M., Azam, M. and Jun, C. (2014).
#'  New Attributes and Variables Control Charts under Repetitive Sampling.
#'  \emph{Industrial Engineering & Management Systems}.
#'  \strong{13}(1):101-106.
#'
#' @importFrom dplyr filter select
#' @importFrom magrittr %>%
#' @importFrom stats dbinom runif
#' @importFrom tibble tibble
#'
#' @export
#'
#' @examples
#'
#' library(magrittr)
#'  npcrs2(
#'   .n     = 40
#' , .p0    = 0.10
#' , .f     = 0.1
#' , .ssize = 1000
#' , .k1r   = 4
#' , .k2r   = .95
#'    )
#'
#'
#' npcrs2(
#'   .n     = 40
#' , .p0    = 0.10
#' , .f     = 0.1
#' , .k1    = 3.13
#' , .k2    = .731
#'    )
#'
#'
#'
#'
npcrs2 <- function(.n, .p0, .f , .ssize = NULL, .k1 = NULL, .k2 = NULL, .k1r = NULL, .k2r = NULL){
  UseMethod("npcrs2")
}

#' @export
#' @rdname npcrs2



npcrs2.default <- function(.n, .p0, .f , .ssize = NULL, .k1 = NULL, .k2 = NULL, .k1r = NULL, .k2r = NULL){
  if(!is.null(.k1) & !is.null(.k2) & is.null(.ssize) & is.null(.k1r) & is.null(.k2r))
  {
    df2 <-  tibble::tibble(
      .f    = .f  
    ,.p1    = .p0 + .f*.p0
    , mu    = .n*.p0
    , sigma = sqrt(.n*.p0*(1-.p0))
    , LCL1  = as.integer(mu - .k1*sigma)
    , UCL1  = as.integer(mu + .k1*sigma)
    , LCL2  = as.integer(mu - .k2*sigma)
    , UCL2  = as.integer(mu + .k2*sigma)
    , Pc    = sum(dbinom(x = (LCL2+1):(UCL2), size = .n, prob = .p0, log = FALSE))
    , Pcc   = sum(dbinom(x = (LCL2+1):(UCL2), size = .n, prob = .p1, log = FALSE))
    , Prep0 = sum(dbinom(x = (LCL1+1):(LCL2), size = .n, prob = .p0, log = FALSE)) + 
              sum(dbinom(x = (UCL2+1):(UCL1), size = .n, prob = .p0, log = FALSE))
    , Prep1 = sum(dbinom(x = (LCL1+1):(LCL2), size = .n, prob = .p1, log = FALSE)) +
              sum(dbinom(x = (UCL2+1):(UCL1), size = .n, prob = .p1, log = FALSE))
    , Pin0  = Pc/(1-Prep0)
    , Pin1  = Pcc/(1-Prep1)
    , ARL0  =   1/(1-Pin0)
    , ARL1  =   1/(1-Pin1)
    , ASN0  =   .n/(1-Prep0)
    , ASN1  =   .n/(1-Prep1)
    ) %>%
      dplyr::select(.f, ASN1, ARL1)
    return(df2)
  }
  else 
    {
     df1 <- 
  tibble::tibble(
    mu    =  .n*.p0
  , sigma =  sqrt(.n*.p0*(1-.p0))
  , k1    = runif(n =  .ssize, min = 3, max = .k1r)
  , k2    = runif(n =  .ssize, min = 0, max = .k2r)  
  ) %>%
  dplyr::filter(k1 > k2) %>%
  dplyr::mutate(
      LCL1 = mu - k1*sigma
    , UCL1 = mu + k1*sigma
    , LCL2 = mu - k2*sigma
    , UCL2 = mu + k2*sigma
    , LCL11 = dplyr::if_else(condition = LCL1 < 0, true = -1, false = LCL1, missing = NULL)
    , LCL22 = dplyr::if_else(condition = LCL2 < 0, true = -1, false = LCL2, missing = NULL)  
  ) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
      Pc   = sum(dbinom(x = as.integer((LCL22 + 1)):as.integer(UCL2 ), size = .n, prob = .p0, log = FALSE))
    , Prep = sum(dbinom(x = as.integer((LCL11 + 1)):as.integer(LCL22), size = .n, prob = .p0, log = FALSE)) + 
             sum(dbinom(x = as.integer((UCL2  + 1)):as.integer(UCL1 ), size = .n, prob = .p0, log = FALSE))
    , Pin0 = Pc/(1 - Prep)
    , ARL0 = 1/(1-Pin0)
    , ASS0 =  .n/(1 - Prep)
  ) %>%
  dplyr::filter(min(ASS0) & ARL0 >= 90) %>%
  dplyr::select(k1, k2, ARL0) %>%
  dplyr::arrange(ARL0) %>% 
  dplyr::filter(dplyr::row_number()==1)
  .k1   <- df1$k1
  .k2   <- df1$k2
  .ARL0 <- df1$ARL0
  mu    <-  .n*.p0
  sigma <-  sqrt(.n*.p0*(1-.p0))
  .f     <- .f
  .p1    <- .p0 + .f*.p0
  LCL11  <- as.integer(mu - .k1*sigma)
  UCL11  <- as.integer(mu + .k1*sigma)
  LCL21  <- as.integer(mu - .k2*sigma)
  UCL21  <- as.integer(mu + .k2*sigma)
  Pc     <- sum(dbinom(x = (LCL21+1):(UCL21), size = .n, prob = .p0, log = FALSE))
  Pcc    <- sum(dbinom(x = (LCL21+1):(UCL21), size = .n, prob = .p1, log = FALSE))
  Prep0  <- sum(dbinom(x = (LCL11+1):(LCL21), size = .n, prob = .p0, log = FALSE)) + 
          sum(dbinom(x = (UCL21+1):(UCL11), size = .n, prob = .p0, log = FALSE))
  Prep1 <- sum(dbinom(x = (LCL11+1):(LCL21), size = .n, prob = .p1, log = FALSE)) +
          sum(dbinom(x = (UCL21+1):(UCL11), size = .n, prob = .p1, log = FALSE))
  Pin0  <- Pc/(1-Prep0)
  Pin1  <- Pcc/(1-Prep1)
  ARL0  <-   1/(1-Pin0)
  ARL1  <-   1/(1-Pin1)
  ASN0  <-   .n/(1-Prep0)
  ASN1  <-   .n/(1-Prep1)
  df2   <- tibble::tibble(.k1, .k2, .f, ASN1, ARL1)
  return(df2)
  }
}


R/qccrs-package.R
#' @name qccrs
#' @title Quality Control Charts under Repetitive Sampling
#' @docType package
#' @description
#'
#' The \code{qccrs} package provides functionalities to calculate
#' Average Sample Numbers (ASN), Average Run Length (ARL1) and 
#' value of k, k1 and k2 for quality control charts under 
#' repetitive sampling as given in Aslam et al. (2014).
#'
#'
#' @author
#' \enumerate{
#'          \item Muhammad Yaseen (\email{myaseen208@@gmail.com})
#'          \item Muhammad Aslam (\email{aslam_ravian@@hotmail.com})
#'          \item Sami Ullah (\email{samiullahuos@@gmail.com})
#'          \item Muhammad Azam   (\email{mazam@@uvas.edu.pk})
#'          \item Chi-Hyuck Jun   (\email{chjun@@postech.ac.kr})
#'          \item Muhammad Kashif (\email{mkashif@@uaf.edu.pk})
#'          }
#'
#' @references
#' Aslam, M., Azam, M. and Jun, C. (2014).
#'  New Attributes and Variables Control Charts under Repetitive Sampling.
#'  \emph{Industrial Engineering & Management Systems}.
#'  \strong{1}(13):101-106.
NULL


R/npcrs1.R
#' @name    npcrs1
#' @aliases npcrs1
#' @title   NP Control Charts under Repetitive Sampling with single positive integer.
#' @description Calculates Average Sample Numbers (ASN), Average Run Length (ARL1) and value of k for NP control charts under repetitive sampling as given in Aslam et al.(2014)
#'
#' @param .n      Sample Size
#' @param .p0     probability that process is in control
#' @param .f      Size of the Shift
#' @param .ssize  Number of samples with replacement at each iteration
#' @param .kr     Random Positive Constant
#' @param .k      Positive Constant
#'
#'
#'
#' @return ARL0, ARL1 and K
#'
#' @author
#' \enumerate{
#'          \item Muhammad Yaseen (\email{myaseen208@@gmail.com})
#'          \item Muhammad Aslam  (\email{aslam_ravian@@hotmail.com})
#'          \item Sami Ullah      (\email{samiullahuos@@gmail.com})
#'          \item Muhammad Azam   (\email{mazam@@uvas.edu.pk})
#'          \item Chi-Hyuck Jun   (\email{chjun@@postech.ac.kr})
#'          \item Muhammad Kashif (\email{mkashif@@uaf.edu.pk})
#'          }
#'
#' @references
#' Aslam, M., Azam, M. and Jun, C. (2014).
#'  New Attributes and Variables Control Charts under Repetitive Sampling.
#'  \emph{Industrial Engineering & Management Systems}.
#'  \strong{13}(1):101-106.
#'
#' @importFrom dplyr filter select
#' @importFrom magrittr %>%
#' @importFrom stats dbinom runif
#' @importFrom tibble tibble
#'
#' @export
#'
#' @examples
#' 
#' library(magrittr)
#' npcrs1(
#'   .n     = 60
#' , .p0    = 0.10
#' , .f     = 0.10
#' , .k     = 2.6432
#' )
#' 
#' 
#' npcrs1(
#'   .n     = 60
#' , .p0    = 0.10
#' , .f     = 0.10
#' , .ssize = 1000
#' , .kr    = 4
#' )
#'
#' 
if(getRversion() >= "2.15.1"){
  utils::globalVariables(
    c(
         "ARL1"
        , "ASN"
        , "ASS0"
	, "LCL1"
	, "LCL2"
	, "LCL22"
	, "Pin"
	, "Pinc"
	, "Prep"
	, "Prep1"
	, "UCL1"
	, "UCL2"
	, "k"
	, "k1"
	, "k2"
    )
  )
}
     
npcrs1 <- function(.n, .p0, .f , .ssize = NULL, .k = NULL,  .kr = NULL){
  UseMethod("npcrs1")
}
#' @export
#' @rdname npcrs1

npcrs1.default <- function(.n, .p0, .f , .ssize = NULL, .k = NULL,  .kr = NULL){
  if(!is.null(.k) & is.null(.ssize) & is.null(.kr))
  {
    mu    <- .n*.p0
    sigma <- sqrt(.n*.p0*(1-.p0))
    .p1   <- .p0 + .f*.p0
    LCL   <- as.integer(mu - .k*sigma)
    UCL   <- as.integer(mu + .k*sigma)
    Pc    <- sum(x = dbinom((LCL+1):UCL, size = .n, prob = .p0, log = FALSE))
    Pcc   <- sum(x = dbinom((LCL+1):UCL, size = .n, prob = .p1, log = FALSE))
    ARL0  <- 1/(1-Pc)
    ARL1  <- 1/(1-Pcc)
    return(tibble::tibble(ARL0 = ARL0 , ARL1=ARL1))
    }  
  else 
    {
      mu    <- .n*.p0
      sigma <- sqrt(.n*.p0*(1-.p0))
      df1 <- 
        tibble::tibble(
          k = runif(n =  .ssize, min = 0, max = .kr)  
        ) %>%
        dplyr::mutate(
            LCL = as.integer(mu - k*sigma)
          , UCL = as.integer(mu + k*sigma)
          , LCL1 = ifelse(test = LCL < 0, yes = -1, no = LCL)
        ) %>%
        dplyr::rowwise() %>%
        dplyr::mutate(
          Pc   = sum(dbinom(x = (LCL1  + 1):UCL, size = .n, prob = .p0, log = FALSE))
          , ARL0 = 1/(1-Pc)
        ) %>%
        dplyr::filter(ARL0 >= 90) %>%
        dplyr::arrange(ARL0) %>%
        dplyr::select(k, ARL0) %>% 
        dplyr::filter(dplyr::row_number()==1)
      
      .k1   <- df1$k
      ARL0  <- df1$ARL0
      mu    <- .n*.p0
      sigma <- sqrt(.n*.p0*(1-.p0))
      .p1   <- .p0 + .f*.p0
      LCL1   <- as.integer(mu - .k1*sigma)
      UCL1   <- as.integer(mu + .k1*sigma)
      Pc1    <- sum(x = dbinom((LCL1+1):UCL1, size = .n, prob = .p0, log = FALSE))
      Pcc1   <- sum(x = dbinom((LCL1+1):UCL1, size = .n, prob = .p1, log = FALSE))
      ARL00  <- 1/(1-Pc1)
      ARL10  <- 1/(1-Pcc1)
      return(tibble::tibble(k = .k1, ARL0 = ARL00 , ARL1=ARL10))
    }
  }
